ros/angel_system_nodes/angel_system_nodes/asr.py

```diff
@@ -0,0 +1,136 @@
+import pyaudio
+import rclpy
+from rclpy.node import Node
+import requests
+import simpleaudio as sa
+import threading
+import time
+import wave
+
+from angel_msgs.msg import HeadsetAudioData, Utterance
+
+import nltk
+from nltk.tokenize import sent_tokenize
+nltk.download('punkt')
+
+
+class ASR(Node):
+    def __init__(self):
+        super().__init__(self.__class__.__name__)
+
+        self.declare_parameter("audio_topic", "HeadsetAudioData")
+        self.declare_parameter("utterances_topic", "utterances_topic")
+        self.declare_parameter("asr_server_url", "http://communication.cs.columbia.edu:8058/home")
+        self.declare_parameter("tmp_wav_filename", "/tmp/asr.wav")
+        
+
+        self._audio_topic = self.get_parameter("audio_topic").get_parameter_value().string_value
+        self._utterances_topic = self.get_parameter("utterances_topic").get_parameter_value().string_value
+        self._asr_server_url = self.get_parameter("asr_server_url").get_parameter_value().string_value
+        self._tmp_wav_filename = self.get_parameter("tmp_wav_filename").get_parameter_value().string_value
+
+        self.subscription = self.create_subscription(
+            HeadsetAudioData,
+            self._audio_topic,
+            self.listener_callback,
+            1)
+        
+        # TODO(derekahmed): Add internal queueing to reduce publisher queue
+        # size to 1.
+        self._publisher = self.create_publisher(
+            Utterance,
+            self._utterances_topic,
+            10
+        )
+
+        self.audio_stream = bytearray()
+        self.audio_stream_lock = threading.RLock()
+        self.asr_server_lock = threading.RLock()
+        self.t = threading.Thread()
+
+        self.audio_stream_duration = 0.0
+        self.segmentation_duration = 30.0
```

Comment:
Could make this is a ROS parameter we pass into this node.

Reply:
Will resolve.